Composer/packages/client/src/telemetry/TelemetryClient.ts

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { LogData, TelemetryEventName, TelemetryEvents, TelemetrySettings } from '@bfc/shared';
+
+import AppInsightsClient from './AppInsightsClient';
+import ConsoleClient from './ConsoleClient';
+
+export default class TelemetryClient {
+  private static _getProperties?: () => LogData;
+  private static _telemetrySettings?: TelemetrySettings;
+
+  public static setup(telemetrySettings: TelemetrySettings, additionalProperties: LogData | (() => LogData)) {
+    if (this._telemetrySettings?.allowDataCollection !== telemetrySettings.allowDataCollection) {
+      this.client?.drain();
+    }
+
+    this._getProperties =
+      typeof additionalProperties === 'function' ? additionalProperties : () => additionalProperties;
+    this._telemetrySettings = telemetrySettings;
+  }
+
+  public static log<TN extends TelemetryEventName>(
+    eventName: TN,
+    ...args: TelemetryEvents[TN] extends undefined ? [never?] : [TelemetryEvents[TN]]
```

Comment:
This typing is a little confusing to me because of `args` usage below. It looks like there is only 1 argument expected ever which is an object. Is that correct?

While I think this is some neat Typescript, what is it buying us here?

Reply:
This allows us to strongly type the `eventName` and the required properties for that event, so we don't inadvertently track an event without the required data.

Reply:
Is the spread arg required or can it be something like

```ts
log<TN extends TelemetryEventName>(
    eventName: TN,
    properties?: TelemetryEvents[TN] extends undefined ? never : TelemetryEvents[TN]
)
```

Reply:
Unfortunately, it doesn't work without the spread operator.

Reply:
Actually.. that does work..